### PR TITLE
Feature/remove shared ptr

### PR DESCRIPTION
**Pull request by reyoung**
This PR is the second step of issue #3520.
This PR is following PR #3522 . Please Review #3522 First.

### CODE REVIEW

File: paddle/framework/backward.cc [hunk 1]
```diff
@@ -91,7 +89,7 @@ std::shared_ptr<OperatorBase> BackwardRecursive(
   }
 
   // Returned gradient network
-  auto net = std::make_shared<operators::NetOp>();
+  auto net = std::unique_ptr<operators::NetOp>(new operators::NetOp());
```
Comment (Superjom) [inline paddle/framework/backward.cc]: can we have a make_unique helper ?

File: paddle/framework/operator.h [hunk 2]
```diff
@@ -116,10 +112,14 @@ class OperatorBase {
...
 
+  // Return a new operator instance, which is as same as this.
+  // Use unique_ptr to prevent caller forget to delete this pointer.
+  virtual std::unique_ptr<OperatorBase> Clone() const = 0;
+
  protected:
   std::string type_;
   // NOTE: in case of OpGrad, inputs_ contains:
-  // I (Inputs)
+  // I (Inputs)opear
```
Comment (Superjom) [inline paddle/framework/operator.h]: what is "opear" mean?

File: paddle/operators/net_op.h [hunk 3]
```diff
@@ -74,21 +84,42 @@ class NetOp : public framework::OperatorBase {
...
+    }
+    ops_.emplace_back(op);
+  }
+
+  void AddOp(std::unique_ptr<framework::OperatorBase>&& op) {
+    AddOp(op.release(), true);
   }
 
-  void InsertOp(size_t pos, const std::shared_ptr<OperatorBase>& op) {
+  void InsertOp(size_t pos, framework::OperatorBase* op, bool own) {
```
Comment (Superjom) [inline paddle/operators/net_op.h]: what's the case when a net does't own an inserted op?

File: paddle/framework/operator.h [hunk 4]
```diff
@@ -130,12 +130,32 @@ class OperatorBase {
...
+#define DEFINE_OP_CLONE_METHOD(CLS)                       \
+  std::unique_ptr<OperatorBase> Clone() const final {     \
+    return std::unique_ptr<OperatorBase>(new CLS(*this)); \
+  }
+
+// Macro for define a default constructor for Operator.
+// You can also use
+//   using PARENT_CLASS::PARENT_CLASS;
+// to use parent's constructor.
+#define DEFINE_OP_CONSTRUCTOR(CLS, PARENT_CLS)                                 \
```
Comment (jacquesqiao) [inline paddle/framework/operator.h]: 这里的命名同上，我看了下caffe2和tensorflow的代码，都是使用了  xx_yy这种方法命名宏的参数，所以要不咱们也按照这个规范来？

File: paddle/framework/operator.h [hunk 5]
```diff
@@ -130,12 +130,32 @@ class OperatorBase {
...
+      const VarNameMap& outputs, const paddle::framework::AttributeMap& attrs) \
+      : PARENT_CLS(type, inputs, outputs, attrs) {}
+
 class NOP : public OperatorBase {
  public:
   using OperatorBase::OperatorBase;
   void InferShape(const Scope& scope) const override {}
   void Run(const Scope& scope,
            const platform::DeviceContext& dev_ctx) const override {}
+  std::unique_ptr<OperatorBase> Clone() const override {
```
Comment (jacquesqiao) [inline paddle/framework/operator.h]: 这里为啥没有统一使用DEFINE_OP_CLONE_METHOD?

File: paddle/framework/backward.cc (hunk 1 above)
Comment (reyoung) [inline paddle/framework/backward.cc]: Well, it could be in another PR. `make_unique` is not in `C++ 11` but in `C++ 14`. We can implement it in `C++ 11`.

File: paddle/framework/operator.h [hunk 6]
```diff
@@ -130,12 +130,32 @@ class OperatorBase {
   AttributeMap attrs_;
 };
 
+// Macro for define a clone method.
+// If you are writing an kernel operator, `Clone` will be defined when you
+// register it. i.e. `Clone` method is not needed to define by yourself.
+#define DEFINE_OP_CLONE_METHOD(CLS)                       \
```
Comment (reyoung) [inline paddle/framework/operator.h]: 那是因为registry里面有很多class同时在一个macro，比如op_class, op_type, grad_class, etc.

File: paddle/framework/operator.h (hunk 2 above)
Comment (reyoung) [inline paddle/framework/operator.h]: Typo here

File: paddle/operators/net_op.h (hunk 3 above)
Comment (reyoung) [inline paddle/operators/net_op.h]: It means that op is owned by another place (other network or owned by Python).

In that situation, the op will be Cloned first.

File: paddle/framework/operator.h (hunk 5 above)
Comment (reyoung) [inline paddle/framework/operator.h]: `DEFINE_OP_CLONE_METHOD`是一个只会被最子类使用的macro，这个函数被标记为final了。
这里，有一些类型会继承NOP，所以不能将Clone标记为final

File: paddle/operators/net_op.h (hunk 3 above)
Comment (Superjom) [inline paddle/operators/net_op.h]: In caffe2's design, every net own its ops? If same op is added, just clone/(create new and copy) it.

If an op belongs to python, python will get a unique_ptr to help control the memory. When net want to add a same op, just clone (create new and copy) it simpler?

File: paddle/framework/operator.h (hunk 4 above)
Comment (reyoung) [inline paddle/framework/operator.h]: 这并没有在Google C++ code style中有描述。

File: paddle/operators/net_op.h (hunk 3 above)
Comment (reyoung) [inline paddle/operators/net_op.h]: Every NetOp own operators in this design. The `own` parameter actually means is this pointer could be moved into NetOp and NetOp will take care that pointer's life cycle.

Maybe that API is confusing. It is removed now.
